### ISSUE TITLE
Moving setStyle back to MapView

### ISSUE THREAD

**Issue by tobrun** [mapbox/mapbox-gl-native#4166]
@bleege @zugaldia 

**TL;DR**:
  - We need to invoke `OnMapReadyCallback` after map has finished loading. 
  - a map can only be loaded if we have set the related style url.
  - `mMapview.setStyle/setStyleUrl` must be called before `mMapview.onCreate` 
  - An exception similar to `InvalidAccessTokenException` will be thrown if not

With the introduction of the `OnMapReadyCallback` we have introduced a async callback that is invoked when we the Map is ready to interact with. Recently I started to notice small side effects when executing methods immediately after this callback has occurred (one strange side effect documented in #4074). It seems that some methods need to be executed before we can interact with the map. Our current implementation is too "quick".

Current implementation just post a message on the messagequeu:

```java
    /**
     * Sets a callback object which will be triggered when the {@link MapboxMap} instance is ready to be used.
     *
     * @param callback The callback object that will be triggered when the map is ready to be used.
     */
    @UiThread
    public void getMapAsync(@NonNull final OnMapReadyCallback callback) {
        post(new Runnable() {
            @Override
            public void run() {
                callback.onMapReady(mMapboxMap);
            }
        });
    }
```

I'm now looking into the possibility to invoke the callback when the Map has finished loading as indicated by the `OnMapChangeListener`. The `getMapAsync` method will look like:

```java

    /**
     * Sets a callback object which will be triggered when the {@link MapboxMap} instance is ready to be used.
     *
     * @param callback The callback object that will be triggered when the map is ready to be used.
     */
    @UiThread
    public void getMapAsync(@NonNull final OnMapReadyCallback callback) {
        mMapReadyCallback = callback;
    }
```

while the `OnMapReadyCallback` will look like:

```java
        // onCreate of MapView
        addOnMapChangedListener(new OnMapChangedListener() {
            @Override
            public void onMapChanged(@MapChange int change) {
                if (change == DID_FINISH_LOADING_MAP) {
                    reloadIcons();
                    reloadMarkers();
                    adjustTopOffsetPixels();
                    mMapReadyCallback.onMapReady(mMapboxMap);
                }
            }
        });
```

Doing this solves these small side effects since markers, icons and offsets are reloaded before interacting with the map but this introduces an extra requirement for the end developer. The `OnMapChangeListener` above will only be invoked if we `setStyle/setStyleUrl` was called.

We used to do it in `4.0.0-SNAPSHOT`with:

```java
        mMapView = (MapView) findViewById(R.id.cameraMapView);
        mMapView.setAccessToken(ApiAccess.getToken(this));
        mMapView.onCreate(savedInstanceState);
        mMapView.getMapAsync(new OnMapReadyCallback() {
            @Override
            public void onMapReady(@NonNull final MapboxMap mapboxMap) {
                // set a style
                mapboxMap.setStyle(Style.MAPBOX_STREETS);
           }        
```

But will now require the end developer to:

```java
        mMapView = (MapView) findViewById(R.id.cameraMapView);
        mMapView.setAccessToken(ApiAccess.getToken(this));
        mMapView.setStyleUrl(Style.MAPBOX_STREETS);
        mMapView.onCreate(savedInstanceState);
        mMapView.getMapAsync(new OnMapReadyCallback() {
            @Override
            public void onMapReady(@NonNull final MapboxMap mapboxMap) {
               // customise map
            }
```

I want to add a note here that this pattern of setting style with accesstoken has been a `semi-requirement` since the beginning of this SDK. The accesstoken was more visible for the end developer since it threw its own exception. Not setting a style was less visible because the developer did not see the app crashing and just saw a black map. Since setting a styleurl will become a more fixed requirement I'm going to introduce a similar check as with accesstoken and throw an exception.

**Comment by tobrun** [comment]
It seems that the callback  is invoked every time a new style url is set, I updated the `OnMapChangeListener` in the following way:


```java

        // Add annotation deselection listener
        addOnMapChangedListener(new OnMapChangedListener() {
            @Override
            public void onMapChanged(@MapChange int change) {
                if (change == DID_FINISH_LOADING_MAP) {
                    reloadIcons();
                    reloadMarkers();
                    adjustTopOffsetPixels();
                    if(mMapReadyCallback!=null) {
                        mMapReadyCallback.onMapReady(mMapboxMap);
                        mMapReadyCallback = null;
                    }
                }
            }
        });

```

**Comment by tobrun** [comment]
I need to check what impact this has if `getMapAsync` is called much later than we are doing in our examples. Eg we could miss the timing of our map already being loaded but we have not set a callback yet.

**Comment by tobrun** [comment]
Impact on Fragments is something else that should be investigated (MapboxMapSettings will always be a solution to work around if a problem occurs).

**Comment by zugaldia** [comment]
> I'm suggesting doing the same as Google maps, load a default style.

:+1:

**Comment by bleege** [comment]
If the choices are between loading a default style and moving `setStyle` back to MapView then I'd vote for the loading a default style.  MapView should only concern itself with View type functionality as well binding to the NativeMapView.
